File: 02/solution.js
var fs = require("fs"); // imports fs

let input;

function Solution()
{
  input = fs.readFileSync("02/input.txt", "utf-8")
    .replace(/\r/gm, "")
    .split("\n")
    .map(i => i.split(" "));

  console.log(input);

  let totalSafeReports = 0;

  // check every record
  for (let r = 0; r < input.length; r++)
  {
    let currentReportSafe = true;
    let 上下 = "unknown";

    // confirm levels are safe or unsafe
    for (let l = 0; l < input[r].length - 1; l++)
    {
      // validate difference between values
      if (input[r][l] === input[r][l + 1] || Math.abs(input[r][l] - input[r][l+1]) > 3)
      {
        currentReportSafe = false;
        break;
      }

      // l and l + 1 are ascending
      if (input[r][l] < input[r][l + 1])
      {
        if (上下 === "下")
        {
          currentReportSafe = false;
          break;
        }
        上下 = "上";
      }

      // l and l + 1 are descending
      else if (上下 === "上")
      {
        currentReportSafe = false;
        break;
      }

      else
      {
        上下 = "下";
      }
    }
    console.log(`report ${r} is ${currentReportSafe}`);
    totalSafeReports += currentReportSafe ? 1 : 0;
  }

  // Solution
  console.log(`The total number of safe reports is ${totalSafeReports}`);
}

Solution();

File: 02/solution.go
package main

import (
	"fmt"
	"os"
	"strconv"
	"strings"
	"time"
)

func AbsInteger(a int) int {
	if a < 0 {
		return -a
	}
	return a
}

func ValidateReport(report []int) bool {
	上下 := "unknown"
	for j := 0; j < len(report)-1; j++ {
		if report[j] == report[j+1] || AbsInteger(report[j]-report[j+1]) > 3 {
			return false
		}
		if report[j] < report[j+1] {
			if 上下 == "下" {
				return false
			}
			上下 = "上"
		} else {
			if 上下 == "上" {
				return false
			}
			上下 = "下"
		}
	}
	return true
}

func main() {

	start := time.Now()

	content, err := os.ReadFile("02/input.txt")
	if err != nil {
		fmt.Printf("Failed to read file: %s\n", err)
		return
	}

	// using an "immediately invoked function"
	reportsList := func() [][]int {
		line := strings.ReplaceAll(string(content), "\r", "")
		lines := strings.Split(line, "\n")
		var reports [][]int
		for _, li := range lines {
			strs := strings.Split(li, " ")
			var report []int
			for _, s := range strs {
				num, _ := strconv.Atoi(s)
				report = append(report, num)
			}
			reports = append(reports, report)
		}
		return reports
	}()

	// solving in a loop
	totalSafeReports := 0
	for _, report := range reportsList {
		if ValidateReport(report) {
			totalSafeReports++
		}
	}

	fmt.Printf("Day 2a solution: %d safe reports\n", totalSafeReports)
	elapsed := time.Since(start)
	fmt.Printf("Elapsed time: %s\n", elapsed)
}

// execute with "go run 02/solution.go"
